packages/back-end/src/services/features.ts

```diff
@@ -906,3 +923,90 @@ export function sha256(str: string, salt: string): string {
     .update(salt + str)
     .digest("hex");
 }
+
+const fromApiEnvSettingsRulesToFeatureEnvSettingsRules = (
+  feature: FeatureInterface,
+  rules: ApiFeatureEnvSettingsRules
+): FeatureInterface["environmentSettings"][string]["rules"] =>
+  rules.map((r) => {
+    if (r.type === "experiment-ref") {
+      const experimentRule: ExperimentRefRule = {
+        // missing id will be filled in by addIdsToRules
+        id: r.id ?? "",
+        type: r.type,
+        enabled: r.enabled != null ? r.enabled : true,
+        description: r.description ?? "",
+        experimentId: r.experimentId,
+        variations: r.variations.map((v) => ({
+          variationId: v.variationId,
+          value: validateFeatureValue(feature, v.value),
+        })),
+      };
+      return experimentRule;
+    } else if (r.type === "force") {
+      const forceRule: ForceRule = {
+        // missing id will be filled in by addIdsToRules
+        id: r.id ?? "",
+        type: r.type,
+        description: r.description ?? "",
+        value: validateFeatureValue(feature, r.value),
+        condition: r.condition,
+        enabled: r.enabled != null ? r.enabled : true,
+      };
+      return forceRule;
+    }
+    const rolloutRule: RolloutRule = {
+      // missing id will be filled in by addIdsToRules
+      id: r.id ?? "",
+      type: r.type,
+      coverage: r.coverage,
+      description: r.description ?? "",
+      hashAttribute: r.hashAttribute,
+      value: validateFeatureValue(feature, r.value),
+      condition: r.condition,
+      enabled: r.enabled != null ? r.enabled : true,
+    };
+    return rolloutRule;
+  });
+
+export const createInterfaceEnvSettingsFromApiEnvSettings = (
+  feature: FeatureInterface,
+  baseEnvs: Environment[],
+  incomingEnvs: ApiFeatureEnvSettings
+): FeatureInterface["environmentSettings"] =>
+  baseEnvs.reduce(
+    (acc, e) => ({
+      ...acc,
+      [e.id]: {
+        enabled: incomingEnvs?.[e.id]?.enabled ?? !!e.defaultState,
+        rules: incomingEnvs?.[e.id]?.rules
+          ? fromApiEnvSettingsRulesToFeatureEnvSettingsRules(
+              feature,
+              incomingEnvs[e.id].rules
+            )
+          : [],
+      },
+    }),
+    {} as Record<string, FeatureEnvironment>
+  );
+
+export const updateInterfaceEnvSettingsFromApiEnvSettings = (
+  feature: FeatureInterface,
+  incomingEnvs: ApiFeatureEnvSettings
+): FeatureInterface["environmentSettings"] => {
+  const existing = feature.environmentSettings;
+  return Object.keys(incomingEnvs).reduce((acc, k) => {
```

Comment:
What happens if the feature currently has environmentSettings for both `dev` and `production`, but the update call only specifies changes for `production`?  If I'm following the logic right, I think this will delete the existing `dev` settings, which we don't want to do.  Is that right?

Reply:
If a feature has envSettings for both `dev` and `production`, and the update only includes `production`, it will not override `dev`. We use `existing` to hold existing envSettings and use that as the initial value for `acc` in the reduce fn, so we only overwrite what is supplied by the update. Additionally, if the update only includes the `enabled` field for an envSetting (and doesn't specify the rules), we **do not** erase the rules but include the existing set of rules with the new `enabled` value ... Let me know if that makes sense

https://github.com/growthbook/growthbook/blob/66f92e3deee419865ed3dc4828675e1d7d0e8f50/packages/back-end/src/services/features.ts#L998-L1011